go/pkg/cas/upload_test.go

```diff
@@ -75,33 +108,63 @@ func TestFS(t *testing.T) {
 			inputs:              []*UploadInput{{Path: filepath.Join(tmpDir, "medium-dir")}},
 			wantScheduledChecks: []*uploadItem{mediumDirItem, mediumItem},
 		},
+		{
+			desc:                "symlinks-preserved",
+			inputs:              []*UploadInput{{Path: filepath.Join(tmpDir, "with-symlinks"), PreserveSymlinks: true}},
+			wantScheduledChecks: []*uploadItem{aItem, subdirItem, cItem, withSymlinksItemPreserved},
+		},
+		{
+			desc:                "symlinks-not-preserved",
+			inputs:              []*UploadInput{{Path: filepath.Join(tmpDir, "with-symlinks")}},
+			wantScheduledChecks: []*uploadItem{aItem, subdirItem, cItem, withSymlinksItemNotPreserved},
+		},
+		{
+			desc:                "dangling-symlinks-disallow",
+			inputs:              []*UploadInput{{Path: filepath.Join(tmpDir, "with-dangling-symlinks")}},
+			expectedErrContains: "no such file or directory",
+		},
+		{
+			desc:                "dangling-symlinks-allow",
+			inputs:              []*UploadInput{{Path: filepath.Join(tmpDir, "with-dangling-symlink"), PreserveSymlinks: true, AllowDanglingSymlinks: true}},
+			wantScheduledChecks: []*uploadItem{withDanglingSymlinksItem},
+		},
 	}
 
 	for _, tc := range tests {
 		t.Run(tc.desc, func(t *testing.T) {
 			var mu sync.Mutex
-			var scheduledCheckCalls []*uploadItem
+			var gotScheduledChecks []*uploadItem
 
 			client := &Client{
 				ClientConfig: DefaultClientConfig(),
 				testScheduleCheck: func(ctx context.Context, item *uploadItem) error {
 					mu.Lock()
 					defer mu.Unlock()
-					scheduledCheckCalls = append(scheduledCheckCalls, item)
+					gotScheduledChecks = append(gotScheduledChecks, item)
 					return nil
 				},
 			}
 			client.SmallFileThreshold = 5
 			client.LargeFileThreshold = 10
 
-			if _, err := client.Upload(ctx, inputChanFrom(tc.inputs...)); err != nil {
-				t.Fatalf("failed to upload: %s", err)
+			_, err := client.Upload(ctx, inputChanFrom(tc.inputs...))
+			if tc.expectedErrContains != "" {
```

Comment:
Use an actual `error` and check for `os.ErrNotExist` instead of string check -- see comment on https://golang.org/pkg/os/#IsNotExist.

This will also simplify this check:

```
if err != tc.expectedErr { t.Fatalf("error mismatch") }
if err != nil { continue } // Happy path where error matches expected non-nil error
```

Reply:
I like the idea, but it doesn't work because today we are using a mix of `errors` and `github.com/pkg/errors` packages. They aren't really compatible. Also we are using an old version of the latter which doesn't have the `Is` function. We should probably update, but this is out of scope of this CL. Added a TODO.

Reply:
nvm, updated the errors package.